controller/sync.go

```diff
@@ -108,21 +108,24 @@ func (m *appStateManager) SyncAppState(app *v1alpha1.Application, state *v1alpha
 		return
 	}
 
-	if syncOp.Source == nil || (syncOp.Sources != nil && len(syncOp.Sources) > 0) {
-		// normal sync case (where source is taken from app.spec.sources)
+	rollback := syncOp.Source != nil && !app.Spec.HasMultipleSources() ||
+		syncOp.Sources != nil && app.Spec.HasMultipleSources()
+
+	if rollback {
+		// rollback case
 		if app.Spec.HasMultipleSources() {
-			sources = app.Spec.Sources
+			sources = state.Operation.Sync.Sources
 		} else {
-			// normal sync case (where source is taken from app.spec.source)
-			source = app.Spec.GetSource()
+			source = *state.Operation.Sync.Source
 			sources = make([]v1alpha1.ApplicationSource, 0)
 		}
 	} else {
-		// rollback case
+		// normal sync case (where source is taken from app.spec.sources)
 		if app.Spec.HasMultipleSources() {
-			sources = state.Operation.Sync.Sources
+			sources = app.Spec.Sources
 		} else {
-			source = *state.Operation.Sync.Source
+			// normal sync case (where source is taken from app.spec.source)
+			source = app.Spec.GetSource()
 			sources = make([]v1alpha1.ApplicationSource, 0)
 		}
 	}
```

Comment:
I think, we can avoid checking whether the app is Multisource App or not on line 111. 
```suggestion
        rollback := (syncOp.Sources != nil && len(syncOp.Sources) > 0) || syncOp.Source != nil
	if !rollback {
		// normal sync case (where source is taken from app.spec.sources)
		if app.Spec.HasMultipleSources() {
			sources = app.Spec.Sources
		} else {
			// normal sync case (where source is taken from app.spec.source)
			source = app.Spec.GetSource()
		}
	} else {
		// rollback case
		if app.Spec.HasMultipleSources() {
			sources = state.Operation.Sync.Sources
		} else {
			source = *state.Operation.Sync.Source
		}
	}
	if sources == nil {
		sources = make([]v1alpha1.ApplicationSource, 0)
	}
```

Let me know what you think.